odoo/upgrade_code/18.1-01-rename-class.py

```diff
@@ -0,0 +1,178 @@
+import re
+
+
+def upgrade(file_manager):
+    nb = len([1 for file in file_manager if file.path.name.endswith('.py')])
+    i = 0
+    file_manager.print_progress(0, nb)
+
+    for file in file_manager:
+        if not file.path.name.endswith('.py'):
+            continue
+
+        class_name_models = get_class_name_models(file)
+        content = file.content
+        content = update_class_name(content, class_name_models)
+        content = remove_name_and_update_inherit(content)
+        content = class_with_2_blank_lines(content)
+        file.content = content
+
+        i += 1
+        file_manager.print_progress(i, nb)
+
+
+regex_camel_case = re.compile(r'(?<=.)([A-Z][a-z])|(?<=[a-z])([A-Z]+(?![a-z]))')
+
+
+def class_name_to_model_name(classname: str) -> str:
+    return regex_camel_case.sub(r'.\1\2', classname).lower()
+
+
+def model_name_to_class_name(model):
+    return ''.join([
+        part[0].upper() + part[1:]
+        for part in re.split(r'[_.]', model)
+        if part
+    ])
+
+
+def attribute_to_model_name(line):
+    if '(' in line or ',' in line:
+        return
+    if '[' in line:
+        # makes the code idempotent
+        return
+    if '"' not in line and "'" not in line:
+        return
+    if line.startswith('     ') or line.strip().startswith('#'):
+        return
+    if not line.split('=', 1)[0].strip() in ('_name', '_inherit'):
+        return
+    return re.sub(r'''[\]\['"\s\n]+''', '', line.split('=').pop())
+
+
+def get_class_name_models(file):
+    content = file.content
+    class_name_models = {}
+    iter_lines = iter(content.split('\n'))
+    line = next(iter_lines, None)
+    while line is not None:
+        class_name = line.startswith('class ') and ('Model)' in line or '.' in line) and line[6:].split(':')[0].split('(')[0]
```

Comment:
Can you write this using a regex?
Without comments it is hard to grasp what is going on (especially the last condition).